File: commonfamily/src/main/java/com/koenig/communication/messages/FamilyMessage.java
package com.koenig.communication.messages;



import com.example.Message;
import com.koenig.commonModel.Byteable;
import com.koenig.commonModel.Component;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.ByteBuffer;

/**
 * Created by Thomas on 11.01.2017.
 */

public abstract class FamilyMessage implements Message {
    public static final String ServerId = "KOENIGSPUTZ_SERVER_ID";
    public static final String SEPARATOR = ";";
    private static final int VERSION_NUMBER = 1;
    protected Logger logger = LoggerFactory.getLogger(getClass().getSimpleName());
    protected String fromId;
    protected String toId;
    protected Component component;
    protected int version;

    public FamilyMessage(Component component) {
        this.component = component;
        this.version = VERSION_NUMBER;
    }

    public FamilyMessage(int version, Component component, String fromId, String toId) {
        this.component = component;
        this.version = version;
        this.fromId = fromId;
        this.toId = toId;
    }


    public abstract String getName();

    public int getTotalLength()
    {
        int length = 4 + 4 + component.getBytesLength() + Byteable.Companion.getStringLength(getName()) + Byteable.Companion.getStringLength(fromId) + Byteable.Companion.getStringLength(toId)
                + getContentLength();
        return length;
    }

    protected abstract int getContentLength();

    protected abstract void writeContent(ByteBuffer buffer);

    public String getFromId()
    {
        return fromId;
    }

    public void setFromId(String fromId)
    {
        this.fromId = fromId;
    }

    public Component getComponent() {
        return component;
    }

    public int getVersion() {
        return version;
    }

    public String getToId()
    {
        return toId;
    }

    public void setToId(String toId)
    {
        this.toId = toId;
    }

    protected void headerToBuffer(ByteBuffer buffer) {
        buffer.putInt(getTotalLength());
        buffer.putInt(version);
        buffer.put(component.toBytes());
        buffer.put(Byteable.Companion.stringToByte(getName()));
        buffer.put(Byteable.Companion.stringToByte(fromId));
        buffer.put(Byteable.Companion.stringToByte(toId));
    }

    public ByteBuffer getBuffer()
    {
        ByteBuffer buffer = ByteBuffer.allocate(getTotalLength());
        headerToBuffer(buffer);
        writeContent(buffer);
        buffer.flip();
        return buffer;
    }
}

File: commonfamily/src/main/java/com/koenig/communication/messages/FamilyMessage.kt
package com.koenig.communication.messages


import com.example.Message
import com.koenig.commonModel.Byteable
import com.koenig.commonModel.Component
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer
import java.util.*

/**
 * Created by Thomas on 11.01.2017.
 */

abstract class FamilyMessage : Message {
    protected var logger = LoggerFactory.getLogger(javaClass.simpleName)
    lateinit var fromId: String
    lateinit var toId: String
    val messageId: String
    var component: Component
        protected set
    var version: Int = 0
        protected set


    abstract val name: String

    val totalLength: Int
        get() = (4 + 4 + component.bytesLength + Byteable.getStringLength(name) + Byteable.getStringLength(fromId) + Byteable.getStringLength(toId) + Byteable.getStringLength(messageId)
                + contentLength)

    protected abstract val contentLength: Int

    constructor(component: Component, messageId: String = UUID.randomUUID().toString()) {
        this.component = component
        this.version = VERSION_NUMBER
        this.messageId = messageId
    }

    constructor(version: Int, component: Component, fromId: String, toId: String, messageId: String = UUID.randomUUID().toString()) {
        this.component = component
        this.version = version
        this.fromId = fromId
        this.toId = toId
        this.messageId = messageId
    }

    protected abstract fun writeContent(buffer: ByteBuffer)

    protected fun headerToBuffer(buffer: ByteBuffer) {
        buffer.putInt(totalLength)
        buffer.putInt(version)
        buffer.put(component.toBytes())
        buffer.put(Byteable.stringToByte(name))
        buffer.put(Byteable.stringToByte(fromId))
        buffer.put(Byteable.stringToByte(toId))
        buffer.put(Byteable.stringToByte(messageId))
    }

    override fun getBuffer(): ByteBuffer {
        val buffer = ByteBuffer.allocate(totalLength)
        headerToBuffer(buffer)
        writeContent(buffer)
        buffer.flip()
        return buffer
    }

    companion object {
        val ServerId = "KOENIGSPUTZ_SERVER_ID"
        val SEPARATOR = ";"
        private val VERSION_NUMBER = 1
    }
}
